tests/plugins/clipboard/datatransfer.js

```diff
@@ -1252,5 +1252,106 @@ bender.test( {
 		var dataTransfer = new CKEDITOR.plugins.clipboard.dataTransfer();
 
 		arrayAssert.itemsAreSame( [], dataTransfer.getTypes() );
+	},
+
+	// (#4604)
+	'test asynchronous getTypes (in browsers with custom types support)': function() {
+		if ( !CKEDITOR.plugins.clipboard.isCustomDataTypesSupported ) {
+			return assert.ignore();
+		}
+
+		var nativeData = bender.tools.mockNativeDataTransfer(),
+			dataTransfer = new CKEDITOR.plugins.clipboard.dataTransfer( nativeData ),
+			expectedTypes = [ 'hublabubla', 'cke4/custom' ];
+
+		nativeData.types = expectedTypes;
+		dataTransfer.cacheData();
+
+		// When asynchronously accessing native data transfer it is unavailable. Here we simulate it by removing stored types.
+		nativeData.types = [];
+
+		arrayAssert.itemsAreSame( expectedTypes, dataTransfer.getTypes() );
+	},
+
+	// (#4604)
+	'test asynchronous getTypes (in browsers without custom types support)': function() {
+		if ( CKEDITOR.plugins.clipboard.isCustomDataTypesSupported ) {
+			return assert.ignore();
+		}
+
+		var nativeData = bender.tools.mockNativeDataTransfer(),
+			dataTransfer = new CKEDITOR.plugins.clipboard.dataTransfer( nativeData ),
+			expectedTypes = [ 'Text', 'URL' ];
+
+		nativeData.setData( 'Text', 'Test' );
+		dataTransfer.cacheData();
+
+		// When asynchronously accessing native data transfer it is unavailable. Here we simulate it by removing stored types.
```

Comment:
TBH I do not see a need for it for several reasons:

1.  Such behaviour is specced:
  
    > Access to the clipboard is performed using the standard `DataTransfer` methods to mutate the items on a `ClipboardEvent`'s `clipboardData` attribute. One consequence of this is that these clipboard APIs can only access clipboard data in the context of a `ClipboardEvent` handler.
	
	Changing it can be considered breaking and web incompatible. This reason alone guarantee I can sleep well at night without having tests for it 😉.
2. There is newer, [async clipboard API](https://w3c.github.io/clipboard-apis/#async-clipboard-api), which allows access to the clipboard anytime. Therefore there is no need to change the existing behaviour of the sync, event-based API.
3. Last but not least: there is no difference to us if such a behaviour changes. Caching data and getting it from cache instead of native `DataTransfer` will work regardless of the asynchronous access to data.

Reply:
Ok, I can agree with `1`.